spring-integration-core/src/main/java/org/springframework/integration/support/MutableMessageBuilder.java

```diff
@@ -73,6 +78,10 @@ public T getPayload() {
 	 */
 	public static <T> MutableMessageBuilder<T> fromMessage(Message<T> message) {
 		Assert.notNull(message, "message must not be null");
+		if (message instanceof CloneableMessage) {
+			logger.warn("For CloneableMessages consider to use regular MessageBuilder. " +
+					"Otherwise custom Message context may be lost.");
+		}
```

Comment:
See my comment for other builder on the matte.
Plus pay attention that we have here `MutableMessageBuilder` and Java doesn't provide multiple inheritance, therefore for `Mutable` case we should have one more delegating wrapper.

Maybe we drop everything and come back to my original solution? :smile: 

I analyzed `MessageWithThreadState` usage and there is only one case when we need to rebuild it. Only that `applySequence` in the `BroadcastingDispatcher`.